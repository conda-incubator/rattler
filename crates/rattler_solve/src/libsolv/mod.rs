use crate::{SolveError, SolverBackend, SolverProblem};
pub use input::cache_repodata;
use input::{add_repodata_records, add_solv_file, add_virtual_packages};
pub use libc_byte_slice::LibcByteSlice;
use output::get_required_packages;
use rattler_conda_types::RepoDataRecord;
use std::collections::HashMap;
use wrapper::{
    flags::SolverFlag,
    pool::{Pool, Verbosity},
    repo::Repo,
    solve_goal::SolveGoal,
};

mod input;
mod libc_byte_slice;
mod output;
mod wrapper;

/// Represents the information required to load available packages into libsolv for a single channel
/// and platform combination
#[derive(Clone)]
pub struct LibsolvRepoData<'a> {
    /// The actual records after parsing `repodata.json`
    pub records: &'a [RepoDataRecord],

    /// The in-memory .solv file built from the records (if available)
    pub solv_file: Option<&'a LibcByteSlice>,
}

impl LibsolvRepoData<'_> {
    /// Constructs a new `LibsolvRepoData` without a corresponding .solv file
    pub fn from_records(records: &[RepoDataRecord]) -> LibsolvRepoData {
        LibsolvRepoData {
            records,
            solv_file: None,
        }
    }
}

/// A [`SolverBackend`] implemented using the `libsolv` library
pub struct LibsolvBackend;

impl SolverBackend for LibsolvBackend {
    type RepoData<'a> = LibsolvRepoData<'a>;

    fn solve<'a, TAvailablePackagesIterator: Iterator<Item = Self::RepoData<'a>>>(
        &mut self,
        problem: SolverProblem<TAvailablePackagesIterator>,
    ) -> Result<Vec<RepoDataRecord>, SolveError> {
        // Construct a default libsolv pool
        let pool = Pool::default();

        // Setup proper logging for the pool
        pool.set_debug_callback(|msg, flags| {
            tracing::event!(tracing::Level::DEBUG, flags, "{}", msg);
        });
        pool.set_debug_level(Verbosity::Low);

        // Add virtual packages
        let repo = Repo::new(&pool, "virtual_packages");
        add_virtual_packages(&pool, &repo, &problem.virtual_packages)
            .map_err(SolveError::ErrorAddingInstalledPackages)?;

        // Mark the virtual packages as installed.
        pool.set_installed(&repo);

        // Create repos for all channel + platform combinations
        let mut repo_mapping = HashMap::new();
        let mut all_repodata_records = Vec::new();
        for repodata in problem.available_packages {
            if repodata.records.is_empty() {
                continue;
            }

            let channel_name = &repodata.records[0].channel;
            let repo = Repo::new(&pool, channel_name);

            if let Some(solv_file) = repodata.solv_file {
                add_solv_file(&pool, &repo, solv_file);
            } else {
                add_repodata_records(&pool, &repo, repodata.records)
                    .map_err(SolveError::ErrorAddingRepodata)?;
            }

            // Keep our own info about repodata_records
            repo_mapping.insert(repo.id(), repo_mapping.len());
            all_repodata_records.push(repodata.records);

            // We dont want to drop the Repo, its stored in the pool anyway, so just forget it.
            std::mem::forget(repo);
        }

        // Create a special pool for records that are already installed or locked.
        let repo = Repo::new(&pool, "locked");
        let installed_solvables = add_repodata_records(&pool, &repo, &problem.locked_packages)
            .map_err(SolveError::ErrorAddingRepodata)?;

        // Also add the installed records to the repodata
        repo_mapping.insert(repo.id(), repo_mapping.len());
        all_repodata_records.push(&problem.locked_packages);

        // Create a special pool for records that are pinned and cannot be changed.
        let repo = Repo::new(&pool, "pinned");
        let pinned_solvables = add_repodata_records(&pool, &repo, &problem.pinned_packages)
            .map_err(SolveError::ErrorAddingRepodata)?;

        // Also add the installed records to the repodata
        repo_mapping.insert(repo.id(), repo_mapping.len());
        all_repodata_records.push(&problem.pinned_packages);

        // Create datastructures for solving
        pool.create_whatprovides();

        // Add matchspec to the queue
        let mut goal = SolveGoal::default();

        // Favor the currently installed packages
        for favor_solvable in installed_solvables {
            goal.favor(favor_solvable);
        }

        // Lock the currently pinned packages
        for locked_solvable in pinned_solvables {
            goal.lock(locked_solvable);
        }

        // Specify the matchspec requests
        for spec in problem.specs {
            let id = pool.intern_matchspec(&spec);
            goal.install(id, false)
        }

        // Construct a solver and solve the problems in the queue
        let mut solver = pool.create_solver();
        solver.set_flag(SolverFlag::allow_uninstall(), true);
        solver.set_flag(SolverFlag::allow_downgrade(), true);

        let transaction = solver
            .solve(&mut goal)
            .map_err(|_| SolveError::Unsolvable)?;

        let required_records = get_required_packages(
            &pool,
            &repo_mapping,
            &transaction,
            all_repodata_records.as_slice(),
        )
        .map_err(|unsupported_operation_ids| {
            SolveError::UnsupportedOperations(
                unsupported_operation_ids
                    .into_iter()
                    .map(|id| format!("libsolv operation {id}"))
                    .collect(),
            )
        })?;

        Ok(required_records)
    }
}
