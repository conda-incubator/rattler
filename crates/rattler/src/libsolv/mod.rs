use std::ffi::CString;

mod ffi;
mod pool;
mod queue;
mod repo;
mod solvable;
mod solver;
mod transaction;

/// Convenience method to convert from a string reference to a CString
fn c_string<T: AsRef<str>>(str: T) -> CString {
    CString::new(str.as_ref()).expect("should be convertable from string")
}

#[cfg(test)]
mod test {
    use crate::libsolv::ffi::{SOLVER_INSTALL, SOLVER_SOLVABLE_PROVIDES};
    use crate::libsolv::pool::{Intern, Pool};
    use crate::libsolv::queue::Queue;
    use rattler::{ChannelConfig, MatchSpec};

    fn conda_json_path() -> String {
        format!(
            "{}/{}",
            env!("CARGO_MANIFEST_DIR"),
            "resources/channels/conda-forge/linux-64/repodata.json"
        )
    }

    #[test]
    fn test_conda_read_repodata() {
        let json_file = conda_json_path();
        let mut pool = Pool::default();
        let mut repo = pool.create_repo("conda-forge");
        repo.add_conda_json(json_file)
            .expect("could not add repodata to Repo");
    }

    #[test]
    fn test_solve_python() {
        let json_file = conda_json_path();
        let mut pool = Pool::default();
        let mut repo = pool.create_repo("conda-forge");
        repo.add_conda_json(json_file)
            .expect("could not add repodata to Repo");
        // Create datastructures for solving
        pool.create_whatprovides();
        let channel_config = ChannelConfig::default();

        // Creat python as a matchspec
        let matchspec =
            MatchSpec::from_str("python", &channel_config).expect("can't create matchspec");
        // Add matchspec to the queue
        let mut queue = Queue::default();
        let id = matchspec.intern(&mut pool);
        queue.push_id_and_flags(id, (SOLVER_INSTALL | SOLVER_SOLVABLE_PROVIDES) as i32);
        // Solver
        let mut solver = pool.create_solver();
        // solve
        solver.solve(&mut queue).expect("unable to solve");

        let mut transaction = solver.create_transaction();
        let solvable_operations = transaction.get_solvable_operations();
    }
}
