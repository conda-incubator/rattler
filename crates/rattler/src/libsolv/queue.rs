use crate::libsolv::ffi;
use std::marker::PhantomData;
use std::os::raw::c_int;

/// Wrapper for libsolv queue type. This type is used by libsolv in the solver to solve for
/// different conda matchspecs. This is a type-safe implementation that is coupled to a specific Id
/// type
pub struct Queue<T> {
    queue: ffi::Queue,
    // Makes this queue typesafe
    _data: PhantomData<T>,
}

impl<T> Default for Queue<T> {
    fn default() -> Self {
        // Safe because we know for a fact that the queue exists
        unsafe {
            // Create a queue pointer and initialize it
            let mut queue = ffi::Queue {
                elements: std::ptr::null_mut(),
                count: 0,
                alloc: std::ptr::null_mut(),
                left: 0,
            };
            // This initializes some internal libsolv stuff
            ffi::queue_init(&mut queue as *mut ffi::Queue);
            Self {
                queue,
                _data: PhantomData,
            }
        }
    }
}

/// This drop implementation drops the internal libsolv queue
impl<T> Drop for Queue<T> {
    fn drop(&mut self) {
        // Safe because this pointer exists
        unsafe {
            ffi::queue_free(self.as_inner_mut());
        }
    }
}

impl<T> Queue<T> {
    /// Returns the ffi::Queue as a mutable pointer
    pub fn as_inner_mut(&mut self) -> *mut ffi::Queue {
        &mut self.queue as *mut ffi::Queue
    }

    /// Returns the ffi::Queue as a const pointer
    pub fn as_inner_ptr(&self) -> *const ffi::Queue {
        &self.queue as *const ffi::Queue
    }
}

impl<T: Into<ffi::Id>> Queue<T> {
    /// Pushes a single id to the back of the queue
    pub fn push_id(&mut self, id: T) {
        unsafe {
            ffi::queue_insert(self.as_inner_mut(), self.queue.count, id.into());
        }
    }

    /// Push multiple id's into the queue
    pub fn push_id_and_flags(&mut self, id: T, flags: i32) {
        unsafe {
            ffi::queue_insert2(
                self.as_inner_mut(),
                self.queue.count,
                flags as c_int,
                id.into(),
            );
        }
    }
}

#[cfg(test)]
mod test {
    use crate::libsolv::pool::StringId;
    use crate::libsolv::queue::Queue;

    #[test]
    fn create_queue() {
        let _queue = Queue::<StringId>::default();
    }
}
