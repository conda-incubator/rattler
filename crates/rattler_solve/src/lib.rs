#![deny(missing_docs)]

//! `rattler_solve` is a crate that provides functionality to solve Conda environments. It currently
//! exposes the functionality through the [`SolverBackend::solve`] function.

mod libsolv;
mod solver_backend;

pub use libsolv::LibsolvBackend;
pub use solver_backend::SolverBackend;
use std::ffi::NulError;

use rattler_conda_types::GenericVirtualPackage;
use rattler_conda_types::{MatchSpec, RepoDataRecord};

/// Represents an error when solving the dependencies for a given environment
#[derive(thiserror::Error, Debug)]
pub enum SolveError {
    /// There is no set of dependencies that satisfies the requirements
    #[error("unsolvable")]
    Unsolvable,

    /// An error occurred when trying to load the channel and platform's `repodata.json`
    #[error("error adding repodata: {0}")]
    ErrorAddingRepodata(#[source] NulError),

    /// An error occurred when trying to load information about installed packages to the solver
    #[error("error adding installed packages: {0}")]
    ErrorAddingInstalledPackages(#[source] NulError),

    /// The solver backend returned operations that we dont know how to install.
    /// Each string is a somewhat user-friendly representation of which operation was not recognized
    /// and can be used for error reporting
    #[error("unsupported operations")]
    UnsupportedOperations(Vec<String>),
}

/// Represents a dependency resolution problem, to be solved by one of the backends (currently only
/// libsolv is supported)
#[derive(Default)]
pub struct SolverProblem {
    /// All available packages
    pub available_packages: Vec<Vec<RepoDataRecord>>,

    /// Records of packages that are previously selected.
    ///
    /// If the solver encounters multiple variants of a single package (identified by its name), it
    /// will sort the records and select the best possible version. However, if there exists a
    /// locked version it will prefer that variant instead. This is useful to reduce the number of
    /// packages that are updated when installing new packages.
    ///
    /// Usually you add the currently installed packages or packages from a lock-file here.
    pub locked_packages: Vec<RepoDataRecord>,

    /// Records of packages that are previously selected and CANNOT be changed.
    ///
    /// If the solver encounters multiple variants of a single package (identified by its name), it
    /// will sort the records and select the best possible version. However, if there is a variant
    /// available in the `pinned_packages` field it will always select that version no matter what
    /// even if that means other packages have to be downgraded.
    pub pinned_packages: Vec<RepoDataRecord>,

    /// Virtual packages considered active
    pub virtual_packages: Vec<GenericVirtualPackage>,

    /// The specs we want to solve
    pub specs: Vec<MatchSpec>,
}

#[cfg(test)]
mod test_libsolv {
    use crate::libsolv::LibsolvBackend;
    use crate::{SolveError, SolverBackend, SolverProblem};
    use rattler_conda_types::GenericVirtualPackage;
    use rattler_conda_types::{
        Channel, ChannelConfig, MatchSpec, NoArchType, PackageRecord, RepoData, RepoDataRecord,
        Version,
    };
    use std::str::FromStr;
    use url::Url;

    fn conda_json_path() -> String {
        format!(
            "{}/{}",
            env!("CARGO_MANIFEST_DIR"),
            "../../test-data/channels/conda-forge/linux-64/repodata.json"
        )
    }

    fn conda_json_path_noarch() -> String {
        format!(
            "{}/{}",
            env!("CARGO_MANIFEST_DIR"),
            "../../test-data/channels/conda-forge/noarch/repodata.json"
        )
    }

    fn dummy_channel_json_path() -> String {
        format!(
            "{}/{}",
            env!("CARGO_MANIFEST_DIR"),
            "../../test-data/channels/dummy/linux-64/repodata.json"
        )
    }

    fn dummy_md5_hash() -> &'static str {
        "b3af409bb8423187c75e6c7f5b683908"
    }

    fn dummy_sha256_hash() -> &'static str {
        "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad"
    }

    fn read_repodata(path: &str) -> Vec<RepoDataRecord> {
        let repo_data: RepoData =
            serde_json::from_str(&std::fs::read_to_string(path).unwrap()).unwrap();
        repo_data.into_repo_data_records(
            &Channel::from_str("conda-forge", &ChannelConfig::default()).unwrap(),
        )
    }

    fn installed_package(
        channel: &str,
        subdir: &str,
        name: &str,
        version: &str,
        build: &str,
        build_number: usize,
    ) -> RepoDataRecord {
        RepoDataRecord {
            url: Url::from_str("http://example.com").unwrap(),
            channel: channel.to_string(),
            file_name: "dummy-filename".to_string(),
            package_record: PackageRecord {
                name: name.to_string(),
                version: Version::from_str(version).unwrap(),
                build: build.to_string(),
                build_number,
                subdir: subdir.to_string(),
                md5: Some(dummy_md5_hash().to_string()),
                sha256: Some(dummy_sha256_hash().to_string()),
                size: None,
                arch: None,
                platform: None,
                depends: Vec::new(),
                constrains: Vec::new(),
                track_features: Vec::new(),
                features: None,
                noarch: NoArchType::default(),
                license: None,
                license_family: None,
                timestamp: None,
            },
        }
    }

    #[test]
    fn test_solve_python() {
        let json_file = conda_json_path();
        let json_file_noarch = conda_json_path_noarch();

        let repo_data = read_repodata(&json_file);
        let repo_data_noarch = read_repodata(&json_file_noarch);

        let available_packages = vec![repo_data, repo_data_noarch];

        let specs = vec![MatchSpec::from_str("python=3.9", &ChannelConfig::default()).unwrap()];

        let problem = SolverProblem {
            available_packages,
            specs,
            ..Default::default()
        };

        let pkgs = LibsolvBackend
            .solve(problem)
            .unwrap()
            .into_iter()
            .map(|pkg| {
                format!(
                    "{} {} {}",
                    pkg.package_record.name, pkg.package_record.version, pkg.package_record.build
                )
            })
            .collect::<Vec<_>>();
        insta::assert_yaml_snapshot!(pkgs);
    }

    #[test]
    fn test_solve_dummy_repo_install_non_existent() {
        let result = solve(
            dummy_channel_json_path(),
            Vec::new(),
            Vec::new(),
            &["asdfasdf", "foo<4"],
        );

        assert!(result.is_err());

        let err = result.err().unwrap();
        assert!(matches!(err, SolveError::Unsolvable));
    }

    #[test]
    fn test_solve_dummy_repo_install_new() -> anyhow::Result<()> {
        let pkgs = solve(
            dummy_channel_json_path(),
            Vec::new(),
            Vec::new(),
            &["foo<4"],
        )?;

        assert_eq!(1, pkgs.len());
        let info = &pkgs[0];

        assert_eq!("foo-3.0.2-py36h1af98f8_1.conda", info.file_name);
        assert_eq!(
            "https://conda.anaconda.org/conda-forge/linux-64/foo-3.0.2-py36h1af98f8_1.conda",
            info.url.to_string()
        );
        assert_eq!("https://conda.anaconda.org/conda-forge/", info.channel);
        assert_eq!("foo", info.package_record.name);
        assert_eq!("linux-64", info.package_record.subdir);
        assert_eq!("3.0.2", info.package_record.version.to_string());
        assert_eq!("py36h1af98f8_1", info.package_record.build);
        assert_eq!(1, info.package_record.build_number);
        assert_eq!(
            "67a63bec3fd3205170eaad532d487595b8aaceb9814d13c6858d7bac3ef24cd4",
            info.package_record.sha256.as_ref().unwrap()
        );
        assert_eq!(
            "fb731d9290f0bcbf3a054665f33ec94f",
            info.package_record.md5.as_ref().unwrap()
        );

        Ok(())
    }

    #[test]
    fn test_solve_dummy_repo_prefers_conda_package() -> anyhow::Result<()> {
        // There following package is provided as .tar.bz and as .conda in repodata.json
        let match_spec = "foo=3.0.2=py36h1af98f8_1";

        let operations = solve(
            dummy_channel_json_path(),
            Vec::new(),
            Vec::new(),
            &[match_spec],
        )?;

        // The .conda entry is selected for installing
        assert_eq!(operations.len(), 1);
        assert_eq!(operations[0].file_name, "foo-3.0.2-py36h1af98f8_1.conda");

        Ok(())
    }

    #[test]
    fn test_solve_dummy_repo_install_noop() -> anyhow::Result<()> {
        let already_installed = vec![installed_package(
            "conda-forge",
            "linux-64",
            "foo",
            "3.0.2",
            "py36h1af98f8_1",
            1,
        )];

        let pkgs = solve(
            dummy_channel_json_path(),
            already_installed,
            Vec::new(),
            &["foo<4"],
        )?;

        assert_eq!(1, pkgs.len());

        // Install
        let info = &pkgs[0];
        assert_eq!("foo", &info.package_record.name);
        assert_eq!("3.0.2", &info.package_record.version.to_string());

        Ok(())
    }

    #[test]
    fn test_solve_dummy_repo_upgrade() -> anyhow::Result<()> {
        let already_installed = vec![installed_package(
            "conda-forge",
            "linux-64",
            "foo",
            "3.0.2",
            "py36h1af98f8_1",
            1,
        )];

        let pkgs = solve(
            dummy_channel_json_path(),
            already_installed,
            Vec::new(),
            &["foo>=4"],
        )?;

        // Install
        let info = &pkgs[0];
        assert_eq!("foo", &info.package_record.name);
        assert_eq!("4.0.2", &info.package_record.version.to_string());

        Ok(())
    }

    #[test]
    fn test_solve_dummy_repo_downgrade() -> anyhow::Result<()> {
        let already_installed = vec![installed_package(
            "conda-forge",
            "linux-64",
            "foo",
            "4.0.2",
            "py36h1af98f8_1",
            1,
        )];

        let pkgs = solve(
            dummy_channel_json_path(),
            already_installed,
            Vec::new(),
            &["foo<4"],
        )?;

        assert_eq!(pkgs.len(), 1);

        // Uninstall
        let info = &pkgs[0];
        assert_eq!("foo", &info.package_record.name);
        assert_eq!("3.0.2", &info.package_record.version.to_string());

        Ok(())
    }

    #[test]
    fn test_solve_dummy_repo_remove() -> anyhow::Result<()> {
        let already_installed = vec![installed_package(
            "conda-forge",
            "linux-64",
            "foo",
            "3.0.2",
            "py36h1af98f8_1",
            1,
        )];

        let pkgs = solve(
            dummy_channel_json_path(),
            already_installed,
            Vec::new(),
            &[],
        )?;

        // Should be no packages!
        assert_eq!(0, pkgs.len());

        Ok(())
    }

    #[test]
    fn test_solve_dummy_repo_with_virtual_package() -> anyhow::Result<()> {
        let pkgs = solve(
            dummy_channel_json_path(),
            Vec::new(),
            vec![GenericVirtualPackage {
                name: "__unix".to_string(),
                version: Version::from_str("0").unwrap(),
                build_string: "0".to_string(),
            }],
            &["bar"],
        )?;

        assert_eq!(pkgs.len(), 1);

        let info = &pkgs[0];
        assert_eq!("bar", &info.package_record.name);
        assert_eq!("1.2.3", &info.package_record.version.to_string());

        Ok(())
    }

    #[test]
    fn test_solve_dummy_repo_missing_virtual_package() {
        let result = solve(dummy_channel_json_path(), Vec::new(), Vec::new(), &["bar"]);

        assert!(matches!(result.err(), Some(SolveError::Unsolvable)));
    }

    #[cfg(test)]
    fn solve(
        repo_path: String,
        installed_packages: Vec<RepoDataRecord>,
        virtual_packages: Vec<GenericVirtualPackage>,
        match_specs: &[&str],
    ) -> Result<Vec<RepoDataRecord>, SolveError> {
        let repo_data = read_repodata(&repo_path);
        let available_packages = vec![repo_data];
        let channel_config = ChannelConfig::default();
        let specs = match_specs
            .into_iter()
            .map(|m| MatchSpec::from_str(m, &channel_config).unwrap())
            .collect();

        let problem = SolverProblem {
            locked_packages: installed_packages,
            virtual_packages,
            available_packages,
            specs,
            ..Default::default()
        };

        let pkgs = LibsolvBackend.solve(problem)?;

        for pkg in pkgs.iter() {
            println!(
                "{} {} {}",
                pkg.package_record.name, pkg.package_record.version, pkg.package_record.build
            )
        }

        if pkgs.len() == 0 {
            println!("No packages in the environment!");
        }

        Ok(pkgs)
    }
}
