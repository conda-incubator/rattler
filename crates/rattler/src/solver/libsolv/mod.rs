use std::ffi::CString;

mod ffi;
mod pool;
mod queue;
mod repo;
mod solvable;
mod solver;
mod transaction;

pub use pool::{Intern, Pool, PoolRef, Verbosity};
pub use queue::Queue;
pub use repo::Repo;
pub use solver::Solver;
pub use transaction::{InstallOperation, Transaction, TransactionRef};

pub use ffi::{SOLVER_INSTALL, SOLVER_SOLVABLE_PROVIDES};

/// Convenience method that converts a string reference to a CString
fn c_string<T: AsRef<str>>(str: T) -> CString {
    CString::new(str.as_ref()).expect("should be convertable from string")
}

#[cfg(test)]
mod test {
    use crate::libsolv::ffi::{SOLVER_INSTALL, SOLVER_SOLVABLE_PROVIDES};
    use crate::libsolv::pool::{Intern, Pool};
    use crate::libsolv::queue::Queue;
    use rattler::{ChannelConfig, MatchSpec};

    use super::pool::Verbosity;

    fn conda_json_path() -> String {
        format!(
            "{}/{}",
            env!("CARGO_MANIFEST_DIR"),
            "resources/channels/conda-forge/linux-64/repodata.json"
        )
    }

    fn conda_json_path_noarch() -> String {
        format!(
            "{}/{}",
            env!("CARGO_MANIFEST_DIR"),
            "resources/channels/conda-forge/noarch/repodata.json"
        )
    }

    #[test]
    fn test_conda_read_repodata() {
        let json_file = conda_json_path();
        let mut pool = Pool::default();
        let mut repo = pool.create_repo("conda-forge");
        repo.add_conda_json(json_file)
            .expect("could not add repodata to Repo");
    }

    #[test]
    fn test_solve_python() {
        let json_file = conda_json_path();
        let json_file_noarch = conda_json_path_noarch();
        let mut pool = Pool::default();
        pool.set_debug_callback(|msg| eprintln!("{}", msg.trim()));
        pool.set_debug_level(Verbosity::Low);
        let mut repo = pool.create_repo("conda-forge");
        repo.add_repodata(
            &serde_json::from_str(&std::fs::read_to_string(json_file).expect("couldnt read"))
                .expect("couldnt parse"),
        )
        .expect("cannot add repodata");
        repo.add_repodata(
            &serde_json::from_str(&std::fs::read_to_string(json_file_noarch).unwrap()).unwrap(),
        )
        .unwrap();
        // repo.add_conda_json(json_file)
        //     .expect("could not add repodata to Repo");
        // repo.add_conda_json(json_file_noarch)
        //     .expect("could not add repodata (noarch) to Repo");
        // Create datastructures for solving
        pool.create_whatprovides();
        let channel_config = ChannelConfig::default();

        // Creat python as a matchspec
        let matchspec =
            MatchSpec::from_str("python=3.9", &channel_config).expect("can't create matchspec");
        // Add matchspec to the queue
        let mut queue = Queue::default();
        let id = matchspec.intern(&mut pool);
        queue.push_id_and_flags(id, (SOLVER_INSTALL | SOLVER_SOLVABLE_PROVIDES) as i32);
        // Solver
        let mut solver = pool.create_solver();
        // solve
        solver.solve(&mut queue).expect("unable to solve");

        let mut transaction = solver.create_transaction();
        let solvable_operations = transaction.get_solvable_operations();
        for operation in solvable_operations.iter() {
            println!(
                "{:?} - {:?}",
                operation.operation,
                operation.solvable.solvable_info()
            );
        }
    }
}
