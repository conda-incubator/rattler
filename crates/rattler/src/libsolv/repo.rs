use std::marker::PhantomData;
use std::ops::{Deref, DerefMut};
use std::ptr::NonNull;

use crate::libsolv::pool::PoolRef;
use crate::libsolv::solvable::SolvableId;
use crate::libsolv::{c_string, ffi};

/// Representation of a repo containing package data in libsolv. This corresponds to a repo_data
/// json. Lifetime of this object is coupled to the Pool on creation
pub struct Repo<'pool>(RepoOwnedPtr, PhantomData<&'pool ffi::Repo>);

#[repr(transparent)]
pub struct RepoRef(ffi::Repo);

impl<'pool> Deref for Repo<'pool> {
    type Target = RepoRef;

    fn deref(&self) -> &Self::Target {
        unsafe { self.0 .0.cast().as_ref() }
    }
}

impl<'pool> DerefMut for Repo<'pool> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        unsafe { self.0 .0.cast().as_mut() }
    }
}

/// Wrapper type so we do not use lifetime in the drop
struct RepoOwnedPtr(NonNull<ffi::Repo>);

/// Destroy c-side of things when repo is dropped
impl Drop for RepoOwnedPtr {
    // Safe because we have coupled Repo lifetime to Pool lifetime
    fn drop(&mut self) {
        unsafe { ffi::repo_free(self.0.as_mut(), 1) }
    }
}

impl RepoRef {
    /// Converts from a pointer to an `ffi::Repo` to a reference of Self.
    pub(super) fn from_ptr<'pool>(repo: NonNull<ffi::Repo>) -> &'pool Self {
        // Safe because a `RepoRef` is a transparent wrapper around `ffi::Repo`
        unsafe { std::mem::transmute(repo.as_ref()) }
    }

    /// Returns a pointer to the wrapped `ffi::Repo`
    fn as_ptr(&self) -> NonNull<ffi::Repo> {
        // Safe because a `RepoRef` is a transparent wrapper around `ffi::Repo`
        unsafe { NonNull::new_unchecked(self as *const Self as *mut Self).cast() }
    }

    /// Returns a reference to the wrapped `ffi::Repo`.
    fn as_ref(&self) -> &ffi::Repo {
        // Safe because a `RepoRef` is a transparent wrapper around `ffi::Repo`
        unsafe { std::mem::transmute(self) }
    }

    /// Returns the pool that created this instance
    pub fn pool(&self) -> &PoolRef {
        // Safe because a `PoolRef` is a wrapper around `ffi::Pool`
        unsafe { &*(self.as_ref().pool as *const PoolRef) }
    }

    /// Reads the content of the file pointed to by `json_path` and adds it to the instance.
    pub fn add_conda_json<T: AsRef<str>>(&mut self, json_path: T) -> anyhow::Result<()> {
        let c_json = c_string(json_path.as_ref());
        let mode = c_string("r");
        unsafe {
            // Cast needed because types do not match in bindgen
            // TODO: see if lib types could be used in bindgen
            // Safe because we check nullptr
            let file = libc::fopen(c_json.as_ptr(), mode.as_ptr()) as *mut ffi::FILE;
            if file.is_null() {
                return Err(anyhow::anyhow!(
                    "fopen returned a nullptr. '{}' does this file exist?",
                    json_path.as_ref()
                ));
            }
            // This line could crash if the json is malformed
            let ret = ffi::repo_add_conda(self.as_ptr().as_mut(), file, 0);
            if ret != 0 {
                return Err(anyhow::anyhow!(
                    "internal libsolv error while adding repodata to libsolv"
                ));
            }

            // Libsolv needs this function to be called so we can work with the repo later
            // TODO: maybe wolf knows more about this function
            ffi::repo_internalize(self.as_ptr().as_mut());
        }
        Ok(())
    }

    /// Adds a `Solvable` to this instance
    pub fn add_solvable(&mut self) -> SolvableId {
        unsafe { SolvableId(ffi::repo_add_solvable(self.as_ptr().as_ptr())) }
    }
}

impl Repo<'_> {
    /// Constructs a new instance
    pub(super) fn new(ptr: NonNull<ffi::Repo>) -> Self {
        Repo(RepoOwnedPtr(ptr), PhantomData::default())
    }
}

#[cfg(test)]
mod tests {
    use crate::libsolv::pool::Pool;

    #[test]
    fn test_repo_creation() {
        let mut pool = Pool::default();
        let mut _repo = pool.create_repo("conda-forge");
    }
}
