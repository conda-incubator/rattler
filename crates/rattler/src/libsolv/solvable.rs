use crate::libsolv::ffi;
use crate::libsolv::pool::{Intern, Pool, StringId};
use crate::libsolv::repo::{Repo, RepoOwnedPtr};
use std::ffi::{CStr, CString};
use std::mem::ManuallyDrop;
use std::ops::DerefMut;
use std::ptr::NonNull;

/// Solvable in libsolv
#[repr(transparent)]
pub struct Solvable(NonNull<ffi::Solvable>);

/// Represents a solvable in a [`Repo`] or [`Pool`]
#[derive(Copy, Clone)]
pub struct SolvableId(pub(super) ffi::Id);

impl From<SolvableId> for ffi::Id {
    fn from(s: SolvableId) -> Self {
        s.0
    }
}

impl SolvableId {
    /// Resolve to the interned type returns a Solvable
    pub fn resolve(&self, pool: &Pool) -> Solvable {
        // Safe because the new-type wraps the ffi::id and cant be created otherwise
        unsafe {
            // Re-implement pool_id2solvable, as it's a static inline function, we can't use it :(
            let solvable = (*pool.0.as_ptr()).solvables;
            // Apparently the solvable is offset by the id from the first solvable
            let solvable = solvable.offset(self.0 as isize);
            Solvable(NonNull::new(solvable).expect("solvable cannot be null"))
        }
    }
}

#[derive(Debug)]
pub struct SolvableInfo {
    name: String,
    version: String,
    build_string: Option<String>,
    build_number: Option<String>,
}

impl Solvable {
    /// Access the inner repo
    pub(super) fn repo(&self) -> ManuallyDrop<RepoOwnedPtr> {
        let repo = unsafe { *self.0.as_ptr() }.repo;
        ManuallyDrop::new(unsafe { std::mem::transmute(repo) })
    }

    /// Returns a solvable info from a solvable
    pub fn solvable_info(&self) -> SolvableInfo {
        let mut pool = self.repo().pool();
        let (name, version, build_string, build_number) = unsafe {
            let solvable = self.0.as_ptr();
            let id = StringId((*solvable).name);
            let version = StringId((*solvable).evr);
            let solvable_build_version = "solvable:buildversion".intern(pool.deref_mut());
            let build_str = ffi::solvable_lookup_str(solvable, solvable_build_version.into());
            let build_string = if !build_str.is_null() {
                Some(
                    CStr::from_ptr(build_str)
                        .to_str()
                        .expect("could not decode string")
                        .to_string(),
                )
            } else {
                None
            };
            let solvable_build_flavor = "solvable:buildflavor".intern(pool.deref_mut());
            let build_number = ffi::solvable_lookup_str(solvable, solvable_build_flavor.into());
            let build_number = if !build_number.is_null() {
                Some(
                    CStr::from_ptr(build_number)
                        .to_str()
                        .expect("could not decode string")
                        .to_string(),
                )
            } else {
                None
            };
            (
                id.resolve(&pool),
                version.resolve(&pool),
                build_string,
                build_number,
            )
        };

        SolvableInfo {
            name: name.to_string(),
            version: name.to_string(),
            build_string,
            build_number,
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::libsolv::pool::Pool;

    #[test]
    pub fn test_solvable_creation() {
        let mut pool = Pool::default();
        let mut repo = pool.create_repo("bla");
        let solvable_id = repo.add_solvable();
        let _solvable = solvable_id.resolve(&pool);
    }
}
