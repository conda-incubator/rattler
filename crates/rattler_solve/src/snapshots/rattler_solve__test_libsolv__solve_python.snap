---
source: crates/rattler_solve/src/lib.rs
assertion_line: 215
expression: pkgs
---
- _libgcc_mutex 0.1 conda_forge
- _openmp_mutex 4.5 1_llvm
- bzip2 1.0.8 h7f98852_4
- ca-certificates 2022.6.15 ha878542_0
- ld_impl_linux-64 2.36.1 hea4e1c9_0
- libffi 3.4.2 h9c3ff4c_2
- libgcc-ng 12.1.0 h8d9b700_16
- libnsl 2.0.0 h7f98852_0
- libstdcxx-ng 12.1.0 ha89aaad_16
- libuuid 2.32.1 h7f98852_1000
- libzlib 1.2.12 h166bdaf_1
- llvm-openmp 14.0.4 he0ac6c6_0
- ncurses 6.3 h9c3ff4c_0
- openssl 1.1.1q h166bdaf_0
- python 3.9.13 h9a8a25e_0_cpython
- readline 8.1.2 h0f457ee_0
- sqlite 3.39.2 h4ff8645_0
- tk 8.6.12 h27826a3_0
- tzdata 2021e he74cb21_0
- xz 5.2.6 h166bdaf_0
- zlib 1.2.12 h166bdaf_1

